src/blocks/subscribe/index.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/**
+ * Newspack Blocks.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack_Newsletters\Blocks\Subscribe;
+
+defined( 'ABSPATH' ) || exit;
+
+const FORM_ACTION = 'newspack_newsletters_subscribe';
+
+/**
+ * Register block from metadata.
+ */
+function register_block() {
+	register_block_type_from_metadata(
+		__DIR__ . '/block.json',
+		array(
+			'render_callback' => __NAMESPACE__ . '\\render_block',
+		)
+	);
+}
+add_action( 'init', __NAMESPACE__ . '\\register_block' );
+
+/**
+ * Enqueue front-end scripts.
+ */
+function enqueue_scripts() {
+	$handle = 'newspack-newsletters-subscribe-block';
+	\wp_enqueue_style(
+		$handle,
+		plugins_url( '../../../dist/subscribeBlock.css', __FILE__ ),
+		[],
+		filemtime( NEWSPACK_NEWSLETTERS_PLUGIN_FILE . 'dist/subscribeBlock.css' )
+	);
+	\wp_enqueue_script(
+		$handle,
+		plugins_url( '../../../dist/subscribeBlock.js', __FILE__ ),
+		[ 'wp-polyfill' ],
+		filemtime( NEWSPACK_NEWSLETTERS_PLUGIN_FILE . 'dist/subscribeBlock.js' ),
+		true
+	);
+	\wp_script_add_data( $handle, 'async', true );
+	\wp_script_add_data( $handle, 'amp-plus', true );
+}
+add_action( 'wp_enqueue_scripts', __NAMESPACE__ . '\\enqueue_scripts' );
+
+/**
+ * Render Registration Block.
+ *
+ * @param array[] $attrs Block attributes.
+ */
+function render_block( $attrs ) {
+	$list_config = \Newspack_Newsletters_Subscription::get_lists_config();
+	if ( empty( $list_config ) ) {
+		return;
+	}
+	$block_id        = \wp_rand( 0, 99999 );
+	$subscribed      = false;
+	$message         = '';
+	$email           = '';
+	$lists           = array_keys( $list_config );
+	$list_map        = array_flip( $lists );
+	$available_lists = array_intersect( $lists, $attrs['lists'] );
+
+	// phpcs:disable WordPress.Security.NonceVerification.Recommended
+	if ( isset( $_REQUEST['newspack_newsletters_subscribed'] ) ) {
+		$subscribed = \absint( $_REQUEST['newspack_newsletters_subscribed'] );
+		if ( isset( $_REQUEST['message'] ) ) {
+			$message = \sanitize_text_field( $_REQUEST['message'] );
+		}
+		if ( isset( $_REQUEST['email'] ) ) {
+			$email = \sanitize_text_field( $_REQUEST['email'] );
+		}
+		if ( isset( $_REQUEST['lists'] ) && is_array( $_REQUEST['lists'] ) ) {
+			$list_map = array_flip( array_map( 'sanitize_text_field', $_REQUEST['lists'] ) );
```

Comment:
In `send_form_response` I only see `newspack_newsletters_subscribed` and `message` parameters set. In what circumstances are `email` and `list` params used?

Reply:
It will be used in case the browser has disabled JS and the backend returns an error. You can see in `send_form_response()` that these parameters are only removed from the redirected response if there are no errors:

```php
		if ( ! $is_error ) {
			$args_to_remove = array_merge( $args_to_remove, [ 'email', 'lists' ] );
		}
```